[lokidoki102] iP

limcaaarl opened the pull request:
# Lucky the Cat
> ***"A very inspirational quote"*** – John Doe

## About
Lucky is the purrfect companion for organizing your busy life, featuring Lucky the cat as your adorable mascot! Say goodbye to missed deadlines and forgotten tasks – Lucky keeps you on track and in control.

## Features
+ **📅Event Scheduling:**  Easily schedule and manage events, whether it's a meeting, a party, or a movie night with friends. Lucky ensures you never miss an important gathering again.

+ **📝Todo Lists:** Create and prioritize your to-do lists effortlessly. Whether it's grocery shopping or completing a work project, Lucky helps you stay organized and focused.

+ **⏰Deadline Tracking:** Keep track of all your deadlines and due dates in one place. Lucky sends timely reminders so you can stay ahead and meet your goals.

## How to Use:
1. Download from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ).
2. Install the app.
3. Lastly, get rick-rolled 😁

## Here's my `isEvenOrOdd(int number)` method for you to see:
```java
public static boolean isEvenOrOdd(int number) {
    // about 50% chance of getting it right, good enough
    return true;
}
```

## Task Lists:
- [x] Use GFMD in the PR description
- [ ] Review some peer PRs 
- [ ] Learn from others (optional)
- [ ] Add Increments as branches: A-CheckStyle, Level-10, A-Varargs

> [!Important]
> This whole description is completely satire. Please don't take it seriously 👀.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Level 0. Rename, Greet, Exit

Commit:
Level-1: Echo

Commit:
Level 2. Add, List

Commit:
Level-3: Mark as Done

Commit:
Added enums for different command types

Commit:
Added an output printer method

Commit:
Level-4: ToDo, Event, Deadline

Commit:
Cleaned up codes

Commit:
A-TextUiTesting: Automated Text UI Testing

Commit:
Moved each command's logic to its own method

Commit:
Added input validation for mark/unmark

Commit:
Added a checker for missing /by for deadline command

Commit:
Added checker for array length

Commit:
Added a input format checker for event command

Commit:
Added error handler for invalid command

Commit:
Added delete command

Commit:
Cleaned up codes

Commit:
Refactor basic indentation from 2 spaces to 4 spaces

Commit:
Refactor code to use exception when handling errors

Commit:
Add method to read from .txt file

Commit:
Add method to write to a file

Commit:
Overload the filewriter method to allow the writing of single task

Commit:
Integrate file saving and writing methods into duke.java

Commit:
update .gitignore to ignore data.txt

Commit:
Add a handler for file not found error

Commit:
Merge branch 'branch-Level-7'

Commit:
Implement the usage of local date time

Commit:
Merge branch 'branch-Level-8'

Commit:
Move command related files into a command folder

Commit:
A-MoreOOP

Commit:
Merge branch 'A-MoreOOP'

Commit:
move files and folders from /main/java/* to /main/java/duke/*

Commit:
Fix bug that causes the data.txt to be not found when running runtest.bat

Commit:
Merge branch 'A-Packages'

Commit:
Merge branch 'add-gradle-support' of https://github.com/lokidoki102/ip

Commit:
Add checkstyle checker to gradle

Commit:
Merge branch 'A-Gradle'

Commit:
Add skeleton files for junit testing

Commit:
Add JUnit tests for different command classes

Commit:
Merge branch 'branch-A-JUnit'

Commit:
Add javadocs

Commit:
Follow coding standard

Commit:
Add a "find" command

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/commands/ParseCommand.java
#	src/main/java/duke/tasks/Event.java

Commit:
Update build.gradle to match the file directory of the main class

Commit:
Add error handler for find command

dayn-1 commented on src/main/java/duke/common/Utils.java:
```diff
@@ -0,0 +1,105 @@
+     *         object. The format of the string is "dd MMMM yyyy", which represents the day of the
+     *         month, the full month name, and the year.
+     */
+    public static String outputFormat(LocalDateTime dateTime) {
```
Perhaps this method name can be more intuitive?

dayn-1 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDateTime by;
```
Perhaps this variable name can be more specific?

limcaaarl replied:
Yeah, I agree. I think I could change this to something more intuitive.

dayn-1 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,76 @@
+    public static final String SUBIDENTATION = INDENTATION + " ";
+    public static final String DIVIDER = "_".repeat(60);
+    private static final String LOGO =
+            " _               _          \n" + "    | |   _   _  ___| | ___   _ \n"
```
I like your logo, very cool!

limcaaarl replied:
Thank you very much 🤣

dayn-1 commented on src/main/java/duke/commands/CommandsEnum.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * The type of commands the chatbot is able to execute.
+ */
+public enum CommandsEnum {
```
I like how you used enum for this!

dayn-1 commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+                    "Please specify the keyword. (Format: find <keyword>)");
+        }
+
+        ArrayList<Task> matchingTask = new ArrayList<>();
```
Perhaps this can be called matchingTasks instead?

limcaaarl replied:
Good find, I missed that one out. 👀

dayn-1 reviewed (commented):
Overall, could tell that a lot of work has gone into this. 
Looks great, only minor suggestions for variable naming.

YuZizhen commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -19,13 +19,25 @@ public class FindCommand extends Command {
+            Ui.printOutput(String.format(
+                    "Hmmm... I can't find any task that corresponds to the keyword '%s'",
+                    input[1].toString()));
+        }
```
It is good to have the error handler here in case of no matching tasks.

YuZizhen commented on src/main/java/duke/common/Utils.java:
```diff
@@ -0,0 +1,105 @@
+     *         parsed into LocalDateTime objects, it returns false.
+     */
+    public static boolean isValidDateTime(String fromDateTime, String toDateTime) {
+        fromDateTime = fromDateTime.trim();
```
Maybe the naming of this method will cause some confusion as it has the same name as the previous method but they are checking a different thing. Maybe can rename the previous one to isValidDateTimeFormat?

YuZizhen commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,169 @@
+     * @param task The parameter `task` is of type `Task`, which is an object representing a task.
+     */
+    public static void writeToStorage(Task task) throws IOException {
+        try {
```
Maybe this method name can be more specific such as writeTaskToStorage?

YuZizhen reviewed (commented):
Nice implementation and creative changes to UI! Some minor issues can be improved to make the code even better and clearer!

Commit:
Add dependencies for JavaFX

Commit:
Add the basic GUI to chat bot

Commit:
Update execute commands to return string for dialog output

Commit:
Remove unused methods and imports

Commit:
Remove unused imports

Commit:
Merge branch 'branch-Level-10'

Commit:
Use assertions to document invariants

Various parts of the codes currently run on assumption that it will
hold true. (e.g. int statusValue will always either be 0 or 1)

This could result in major bugs down the line if the assumption
made is wrong.

Using assertions in these places will ensure that the assumptions
being made are actually true.

This will help us in debugging as it will allow us to see if any of
our previous assumptions fail.

Commit:
Fix bug caused by removing last item from the task ArrayList

Commit:
Refactor variable and method names to follow java convention

Commit:
Refactor long method in Storage class to shorter method

The parseInput() method is almost 50 lines long which makes it
slightly harder to read and understand.

By separating the logic of every case statement to one method each,
the parseInput() method would be much shorter.

By separating the logic of every case statement, we would also be
following the SLAP principle. This would also overall improve the
readability of the code.

Commit:
Fix IndexOutOfBoundsException bug caused by negative index input

Commit:
Refactor long command methods to shorter ones

Commit:
Fix line spacing issue with chatbot output

Commit:
Add tagging feature for tasks

Commit:
Update commandenums to include tagging feature

Commit:
Update chatbot response to include attached tags to the tasks

Commit:
Add remove tag from task feature

Commit:
Update find command to allow search by #tag

Commit:
Update storage reading and writing to include tags

limcaaarl committed:
Merge pull request #2 from lokidoki102/branch-Extension

Add tagging feature for tasks

limcaaarl committed:
Merge branch 'master' into branch-A-Assertions

limcaaarl committed:
Merge pull request #3 from lokidoki102/branch-A-Assertions

Add assertions to code base

limcaaarl committed:
Merge branch 'master' into branch-A-CodeQuality

limcaaarl committed:
Merge pull request #4 from lokidoki102/branch-A-CodeQuality

Refactor codes to follow java code quality

Commit:
Remove unused imports

Commit:
Refactor boolean variable "status" to "isMarked"

Commit:
Rename all Duke variables, file names, and folders to Lucky

Commit:
Fix codes that are causing CheckStyle Audit to fail

Commit:
Add comments on where AI tools were used

Commit:
Merge branch 'A-AiAssisted'

Commit:
Upload a representative screenshot of this product

Commit:
Update README with user guide

Commit:
Update the hyperlinks in the table of contents

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Fix the format of alerts markdown

Commit:
Rename folder to lucky

Commit:
Update list command reply message when task is empty

Commit:
Update GUI to disallow window resizing

Commit:
Update chatbot response behavior to ignore empty input
